cpp/src/io/parquet/stream_pool.cpp

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <mutex>
+
+#include "stream_pool.hpp"
+
+#include <cudf/utilities/error.hpp>
+
+namespace cudf::io::detail::parquet {
+
+namespace {
+
+std::size_t constexpr STREAM_POOL_SIZE = 32;
+
+// doing lazy initialization to avoid allocating this before cuInit is called (was
+// causing issues with compute-sanitizer).
+// deliberately allowing the pool to leak to avoid deleting streams after cuda shutdown.
+rmm::cuda_stream_pool* pool = nullptr;
+
+std::mutex stream_pool_mutex;
+
+void init()
+{
+  std::lock_guard<std::mutex> lock(stream_pool_mutex);
```

Comment:
I think we don't need a mutex at any point (except possibly in get_streams) if the singleton is implemented with a static variable, similar to `logger_wrapper`.

Reply:
I'd be happy to get rid of the `init()` if possible.  Didn't think to try static variable like the old version...I had it defined in the anonymous namespace.